### ISSUE TITLE
macOS 13.4 issues

### ISSUE THREAD

**Issue by zearp** [zearp/Nucintosh#129]
Seems the macOS 13.4 update introduces some issues when using Intel bluetooth/wireless. Bluetooth is completely broken and the Intel wifi kext is preventing proper sleep. Sleep works fine with the Intel kexts disabled but then you have no working bluetooth and wireless.

There are no issues if you disabled Intel bt/wifi and use a natively supported Apple card. I've tested both stock and modded. I'll update here if I find any solutions. For now it seems best to stay on 13.3.x if you rely on Intel bluetooth/wireless and want working sleep without disabling anything.

We'll need to wait for fixes for these issues upstream, they've already been reported by people so the developers are aware.

https://github.com/acidanthera/BrcmPatchRAM
https://github.com/OpenIntelWireless/itlwm
https://github.com/OpenIntelWireless/IntelBluetoothFirmware

**Comment by naryfa** [comment]
Awesome, glad to know. I will still do a full backup before updating. I use the natively (hopefully still) supported BCM94360CS2 so I guess I'm fine?

**Comment by zearp** [comment]
Having an up to date bootable backup with a working EFI is something I would highly recommend indeed! 

Also you could make a bootable clone and test updates in there and if it works either apply the updates again to the main OS or clone the updated backup back.

I'm not sure what you mean about OCLP and cpu stuff. Our 8th gen Intel is supported natively and that won't change soon. I also would not change SMBIOS to anything not 8th gen Intel as it does many things including frequencies and such. It is important the SMBIOS matches the our platform. 

It is possible to use OCLP to force things but it's far from ideal to do that needlessly. For example I tried it on an older Mac and while Ventura works things like power management are far from idea as the SMBIOS is of a much newer platform.

Bluetooth/wireless combo cards that are supported natively will work until Appel removed support and even if they do they usually don't fully remove it and it is possible to patch it back in like I did for the older Apple cards here:

https://github.com/zearp/Nucintosh/blob/master/EFI/OC/config.plist#L550-L582

This no longer works on Ventura for this card but worked fine on Big Sur. It changes the model number in the driver so the driver loads even when you use an SMBIOS that doesn't enable support for that card out of the box.

**Comment by naryfa** [comment]
Great piece of info, thank you. Cloning saved my @$$ so many times. 

The SMBIOS stays, of course. I was just curious if there were any benchmark hits after this update.

**Comment by zearp** [comment]
Things are looking good! The latest commits suggest the delays on Ventura when using Intel wireless have been/are being addressed. I will compile a new version and commit it to the repo tonight after testing.

https://github.com/OpenIntelWireless/itlwm/commits/master

**Comment by zearp** [comment]
> Great piece of info, thank you. Cloning saved my @$$ so many times.

Same here! I really like macOS in that regard, it is easy to clone and boot into the clones. Not so easy with Linux or Windows. macOS also doesn't care what it boots from. On my MacBooks I can boot from sd cards if I wanted to have a bad time 😅

**Comment by zearp** [comment]
I can confirm the wireless connection delay has been fixed upstream and wifi connects a lot faster 🥳

Updated kext is here: https://github.com/zearp/Nucintosh/blob/master/Stuff/AirportItlwm.kext-Ventura.zip

No news on the bluetooth front as of yet. Only some work arounds I rather not apply but I'll test them today or tomorrow and see if we can implement it if no progress is made upstream.

**Comment by zearp** [comment]
I wasn't able to pair any devices using bluetooth both new and previously paired devices do not connect. I'll do some more testing soon.

As a work around I'm using some 99 cent bluetooth 4.0 dongles I got a dozen of while ago from eBay. They can be identified by white lettering saying CSR 4.0 on the plastic.

```Bus 020 Device 004: ID 0a12:0001 Cambridge Silicon Radio, Ltd Bluetooth Dongle (HCI mode)```

They work out of the box, I also seen them on Amazon for about 2-5 euro. They also have newer 5.0 ones but I haven't tested those so I'm not sure if they work natively.

![bt_dongle](https://github.com/zearp/Nucintosh/assets/63272687/714a12b7-c185-465f-aeac-2560ee7c790d)

For the dongle to work properly the usb port kext must be edited; [HS07](https://github.com/zearp/Nucintosh/blob/master/EFI/OC/Kexts/USBPorts.kext/Contents/Info.plist#L150-L158) needs to be removed and bluetooth in the bios should be disabled. The ```IBlueToolFixup.kext``` and ```IntelBluetoothFirmware.kext``` entries should also be disabled in the OpenCore config file.

**Comment by naryfa** [comment]
> > Great piece of info, thank you. Cloning saved my @$$ so many times.
> 
> Same here! I really like macOS in that regard, it is easy to clone and boot into the clones. Not so easy with Linux or Windows. macOS also doesn't care what it boots from. On my MacBooks I can boot from sd cards if I wanted to have a bad time 😅

Mind me asking what do you clone with? Sorry to side-step here.

**Comment by zearp** [comment]
SuperDuper! as its free to use and works great. Just updating an existing backup is where you have to pay but creating a brand new bootable clone is free. Just make a clone then copy the EFI to it and you're good to go. You can also boot into Time Machine backups from OpenCore.

**Comment by aacassandra** [comment]
> I can confirm the wireless connection delay has been fixed upstream and wifi connects a lot faster partying_face
> 
> Updated kext is here: https://github.com/zearp/Nucintosh/blob/master/Stuff/AirportItlwm.kext-Ventura.zip
> 
> No news on the bluetooth front as of yet. Only some work arounds I rather not apply but I'll test them today or tomorrow and see if we can implement it if no progress is made upstream.

how about bluetooth? my bluetooth can't active. im fresh install with macos ventura 13.4

**Comment by zearp** [comment]
@aacassandra Stated in the first post:

>  Bluetooth is completely broken and the Intel wifi kext is preventing proper sleep.

**Comment by aacassandra** [comment]
sorry in advance, what does that mean. I just need to update my wifi kext and it solves the problem?
@zearp

**Comment by aacassandra** [comment]
hi @zearp  , what do you know?, where i can download ventura 13.3 and i can create bootable usb from linux.

**Comment by zearp** [comment]
Download any macOS version with this: https://github.com/corpnewt/gibMacOS

You can not create full installers on any OS other than macOS, you can only create bootable internet installers that will download the 12GB installer during install. I've never did a recovery install nor do I want to support those. Making it in macOS using the official tool is best and you can run a basic macOS in a (K)VM easily to do so.

https://dortania.github.io/OpenCore-Install-Guide/installer-guide/linux-install.html

**Comment by hajabooboo2** [comment]
> I wasn't able to pair any devices using bluetooth both new and previously paired devices do not connect. I'll do some more testing soon.
> 
> As a work around I'm using some 99 cent bluetooth 4.0 dongles I got a dozen of while ago from eBay. They can be identified by white lettering saying CSR 4.0 on the plastic.
> 
> `Bus 020 Device 004: ID 0a12:0001 Cambridge Silicon Radio, Ltd Bluetooth Dongle (HCI mode)`
> 
> They work out of the box, I also seen them on Amazon for about 2-5 euro. They also have newer 5.0 ones but I haven't tested those so I'm not sure if they work natively.
> 
> ![bt_dongle](https://user-images.githubusercontent.com/63272687/242706284-714a12b7-c185-465f-aeac-2560ee7c790d.jpg)
> 
> For the dongle to work properly the usb port kext must be edited; [HS07](https://github.com/zearp/Nucintosh/blob/master/EFI/OC/Kexts/USBPorts.kext/Contents/Info.plist#L150-L158) needs to be removed and bluetooth in the bios should be disabled. Also `BlueToolFixup.kext` should be enabled and `IntelBluetoothFirmware.kext` should be disabled in the OpenCore config file.

First, thanks a lot for sharing your EFI stuff and helpful guides :)

I tried your workaround (and many other options) but had no success getting BT with two dongles (CSR, BCM20702A0) to work using your latest EFI release.

Do you have any advice?

**Comment by hajabooboo2** [comment]
Got the internal BT running with this [FW](https://github.com/Lorys89/IntelBluetoothFirmware/suites/13685209550/artifacts/756385483) and this [BlueToolFixup.kext](https://github.com/acidanthera/BrcmPatchRAM/suites/13641000431/artifacts/752879905).

**Comment by aacassandra** [comment]
> Got the internal BT running with this [FW](https://github.com/Lorys89/IntelBluetoothFirmware/suites/13685209550/artifacts/756385483) and this [BlueToolFixup.kext](https://github.com/acidanthera/BrcmPatchRAM/suites/13641000431/artifacts/752879905).

does it work for you? im still using ventura 13.3.1 hehe

**Comment by hajabooboo2** [comment]
> 

Yeah, in my case it also works for Ventura 13.4.

**Comment by zearp** [comment]
> Got the internal BT running with this [FW](https://github.com/Lorys89/IntelBluetoothFirmware/suites/13685209550/artifacts/756385483) and this [BlueToolFixup.kext](https://github.com/acidanthera/BrcmPatchRAM/suites/13641000431/artifacts/752879905).

Oh so we may just need to update the firmware. That kext has *all* the firmware so I will compile from source with only the firmware we need. I'll test it first and if it works I'll push out a new release hopefully tonight or tomorrow. I'm sure other kexts also have had updates and hopefully 13.4 and 14.x fixes.

**Comment by zearp** [comment]
I've quickly spun up a test EFI and pushed the newly compiled stuff to the repo. This also includes some new commits to the Intel wifi driver but I don't know if it fixes anything for 14.x and the slow connections were solved in the current version.

[EFI.zip](https://github.com/zearp/Nucintosh/files/11835717/EFI.zip)

**Comment by hajabooboo2** [comment]
> I've quickly spun up a test EFI and pushed the newly compiled stuff to the repo. This also includes some new commits to the Intel wifi driver but I don't know if it fixes anything for 14.x, the slow AP connection/scan times on Ventura and earlier were solved in the current version.
> 
> [EFI.zip](https://github.com/zearp/Nucintosh/files/11835717/EFI.zip)
> 
> Please test and let me know!

Quick test: WiFi and BT work ;)

**Comment by zearp** [comment]
Can report the same! Wireless was a bit weird at first when it said my wifi was hidden but it works fine now!

**Comment by zearp** [comment]
Pairing devices is still a bit unpredictable. Let's hope some more fixes appear soon.

**Comment by naryfa** [comment]
I have updated to 13.4.1 on June 24, 2023 and everything seems to be working fine. I do have the original WiFi/BT card though. If anything pops up, I will update this comment.

**Comment by zearp** [comment]
I also noticed some usb weirdness including on my real Macs, it is most likely Apple has changed something in the underlaying drivers.

I've updated everything to the latest available versions, hopefully this will fix some of the issues. It would be a bit of a downer if the recent macOS updates are breaking this pretty much golden EFI.

https://github.com/zearp/Nucintosh/releases/download/v4.4/EFI_04072023.zip

Please try and let me know if it made things better, especially if you're using Intel bt/wifi. I don't notice any difference myself but I mainly have NUCs with Broadcom bt/wifi cards.

**Comment by zearp** [comment]
Clean install of 13.4.1 on a stock NUC went perfectly fine. Everything works so it seems all the issues have been sorted. I'll keep an eye on the usb issues. To me it seems as if the power delivery or something has changed a bit. As I test I connected a bunch of usb mice and keyboard and some just turned off. Seemingly not getting enough power. I don't know enough about the whole usb port thing but I do know it does (try) to control power delivery. I also admit that on my MacBook this is an issue sometimes where plugging in a usb device sometimes stops others. Surely not going above the 15 port limit so my best guess would be power.

Other than the weird usb issues that seem to pop up sometimes all is good on the 13.x front now for both stock and NUCs with Broadcom. Let's hope Sonoma will be smooth sailing once these initial driver issues and such have been fixed. I don't think Apple will officially support Coffee Lake on 15.x macOS but who knows!

**Comment by zearp** [comment]
What is funny is that this Microsoft mouse I use for testing only shows its remaining battery when using Intel bluetooth, whilst the natively supported Broadcom bluetooth doesn't show it.

![Screenshot 2023-07-04 at 18 01 16](https://github.com/zearp/Nucintosh/assets/63272687/7375fc5f-62f8-463d-b5d3-bf1ed7cda0c4)

**Comment by zearp** [comment]
If you have any weird issues after updating consider making a backup (Time Machine or clone) and then do a clean install formatting the disk and use the migration tool to restore all your settings and apps. Using the full installer to install on top of an existing install can also cure many issues. Resetting NVRAM can't hurt as well. Funnily these also apply to real Macs lol.

**Comment by zearp** [comment]
If you have any weird issues after updating consider making a backup (Time Machine or clone) and then do a clean install formatting the disk and use the migration tool to restore all your settings and apps. Using the full installer to install on top of an existing install can also cure many issues. You won't lose any apps or settings. Resetting NVRAM can't hurt as well. Funnily these also apply to real Macs lol.

**Comment by naryfa** [comment]
I noted everything you've said. It's likely I won't do a clean install because it's my main hackintosh and I only have one. No time frame to re-configure everything. I should have some time this coming week to fiddle around, I will reset NVRAM and if that doesn't do anything, I'll run the install on top of the existing one. I'm just super glad that someone else has noted the USB gremlins. 

I just realized I may be using an older EFI because it boots super fast and I completely forgot about it. Anyways, I'll do my tests and if I find anything, I'll report.

The battery level reporting situation is somewhat ironic.

**Comment by zearp** [comment]
I remember having similar usb weirdness on the Dell machines but it went away as mysteriously as it started. That was a few years ago but it had an issue where some devices would be ejected/mounted over and over and over.

It is likely you don't really have to change much in the EFI for it to boot the latest beta 3 but some stuff might not work due to changes in Sonoma, like Intel wireless/bluetooth. The latest release EFI works 100% now on both Ventura and Sonoma. So it is still a "golden" EFI in my eyes.

I haven't found a way to reproduce the usb weirdness but my guess is that it has something to do with power delivery which can be a bit configured in an acpi patch ([SSDT-EC-USBX.aml](https://github.com/zearp/Nucintosh/blob/master/EFI/OC/ACPI/SSDT-EC-USBX.aml)).

```
                Return (Package (0x08)
                {
                    "kUSBSleepPowerSupply", 
                    0x13EC, 
                    "kUSBSleepPortCurrentLimit", 
                    0x0834, 
                    "kUSBWakePowerSupply", 
                    0x13EC, 
                    "kUSBWakePortCurrentLimit", 
                    0x0834
                })
```

This is what we use now, which is the default recommended value in the Dortania guide. The thing is that I can't reproduce it so expimentign with these values is kind of useless until we can trigger the usb weirdness reliably. I will see if there's some kind of way to get whatever values Linux or Windows uses and use those. then again I could be wrong the usb stuff is not caused by a power issue. It is the best guess I got right now.

**Comment by zearp** [comment]
It is likely you don't really have to change much in the EFI for it to boot the latest beta 3 but some stuff might not work due to changes in Sonoma, like Intel wireless/bluetooth. The latest release EFI works 100% now on both Ventura and Sonoma on stock NUCs. So it is still a "golden" EFI in my eyes.

I remember having similar usb weirdness on the Dell machines but it went away as mysteriously as it started. That was a few years ago but it had an issue where some devices would be ejected/mounted over and over and over. I haven't found a way to reproduce the usb weirdness but my guess is that it has something to do with power delivery which can be slightly configured in an acpi patch (SSDT-EC-USBX.aml).

```
                Return (Package (0x08)
                {
                    "kUSBSleepPowerSupply", 
                    0x13EC, 
                    "kUSBSleepPortCurrentLimit", 
                    0x0834, 
                    "kUSBWakePowerSupply", 
                    0x13EC, 
                    "kUSBWakePortCurrentLimit", 
                    0x0834
                })
```
This is what we use now, which are the default recommended values from the Dortania guide. The thing is that I can't reproduce it so experimenting with these values is kind of useless until we can trigger the usb weirdness reliably. I will see if there's some kind of way to get whatever values Linux or Windows uses and use those. Then again I could be wrong the usb stuff is not caused by a power issue. It is the best guess I got right now.

**Comment by naryfa** [comment]
I see, I will use the latest one you released. I thought I was using the old one, but I did actually update to a newer version. I just have to plug in my values and drop Intel stuff. Do you suggest I use the ACPI patch you mentioned?

**Comment by zearp** [comment]
That ACPI patch is already included as it's required. I was just speculating about the usb cause and possible solution of the usb issues that can't seem to be reproduced.

**Comment by naryfa** [comment]
OK. The USB issues usually occurred to me when all of the ports were populated.

**Comment by hac-mac** [comment]
one question regarding sleep - I updated to Bios 94 some weeks ago.
Post that update display won't turn on anymore, but NUC is running.

It was working without any issues before, not sure if I updated the EFI around that as well but I think not.
The HDMI patch was also installed before, something seems to got broke.

Not sure how this can be retrieved via logs. If not waking up I've to unplug power to make it work again

**Comment by zearp** [comment]
@hac-mac Nothing has changed in the EFI regarding that in a very long time. Usually turning force-online on fixes these things but thats already enabled by default. 

https://github.com/zearp/Nucintosh/blob/master/EFI/OC/config.plist#L197-L198C13

Updating the BIOS is something I stopped doing once I figured out Intel just disabled our ability to undervolt and we can't turn it back on even when force downgrading the BIOS. I guess it's part of the ME and that can't be downgraded. So in a way Intel took a function away that kept my NUCs cooler and using less energy. Thank you Intel for not giving us any options to re-enable undervolting. I don't think this is caused by BIOS though, more people would've ran into the same issue.

Good luck!
